File: frontend/turing-machine-web/src/logic/Heads/ReadWriteHead.ts
import { IHead } from "./IHead"
import { HeadTypes } from "./HeadTypes"
import { ITape } from "../Tapes/ITape"

export class ReadWriteHead implements IHead
{
    // Getter for the head type
    public get Type(): HeadTypes {
        return HeadTypes.ReadWrite;
    }

    public Position: number = 0;
    public TapeID: number = 0;

    private readonly _useTape: ITape;

    /**
     * ReadWriteHead constructor accepts a tape object for operations. Cannot change later.
     * Factory method should pass this object by the given tape reference ID.
     * @param tape An ITape object.
     */
    constructor(tape: ITape)
    {
        this._useTape = tape;
    }

    public GetCurrentContent(): string | null
    {
        const { success, content } = this._useTape.TryRead(this.Position);

        if (success)
            return content;
        else
            return null;
    }

    public Move(steps: number): void
    {
        this.Position = this._useTape.GetMovedPosition(this.Position, steps);
    }

    public TryWrite(content: string, machineID: number, headID: number): boolean
    {
        try 
        {
            this._useTape.ScheduleWrite(this.Position, content, machineID, headID);
        }
        catch (ex)
        {
            if (ex instanceof Error) {
                console.log(ex.message); // Safely access the message property
            } else {
                console.log("An unknown error occurred.");
            }

            return false;
        }

        return true;
    }

    public IsUsesTape(tape: ITape): boolean {
        return tape === this._useTape;
    }
}

File: archive/logic/TuringMachineSimulationNew/Heads/ReadWriteHead.cs
using TuringMachineSimulation.Logic.Tapes;

namespace TuringMachineSimulation.Logic.Heads
{
    internal class ReadWriteHead : IHead
    {
        public HeadType Type => HeadType.ReadWrite;

        public int Position { get; set; }

        private readonly ITape _tape;

        public int TapeID { get; set; }

        /// <summary>
        /// ReadWriteHead constructor, accepts a tape object for operations. Cannot change later.
        /// </summary>
        /// <param name="tape"> an ITape object. </param>
        public ReadWriteHead(ITape tape)
        {
            _tape = tape;
        }

        public char? GetCurrentContent()
        {
            _tape.TryRead(Position, out var currentContent);
            return currentContent;
        }
        
        public void Move(int steps)
        { 
            Position = _tape.GetMovedPosition(Position, steps);
        }
        
        public bool TryWrite(char content, int machineID, int headID)
        {
            try
            {
                _tape.ScheduleWrite(Position, content, machineID, headID);
            }
            catch (ArgumentOutOfRangeException)
            {
                Console.WriteLine($"[Error] Position or content is out of range. " +
                                  $"Position: {Position}, Content: {content}.");
                return false;
            }
            catch (InvalidOperationException ex)
            {
                Console.WriteLine(ex.Message);
                return false;
            }
            
            return true;
        }

        public bool IsUsesTape(ITape tape)
        {
            return object.ReferenceEquals(_tape, tape);
        }
    }
}
